File: src/main/java/work/lclpnet/ap2/game/apocalypse_survival/util/MonsterSpawner.java
package work.lclpnet.ap2.game.apocalypse_survival.util;

import net.minecraft.entity.EntityType;
import net.minecraft.entity.SpawnReason;
import net.minecraft.entity.ai.goal.*;
import net.minecraft.entity.ai.pathing.EntityNavigation;
import net.minecraft.entity.ai.pathing.MobNavigation;
import net.minecraft.entity.attribute.EntityAttributeInstance;
import net.minecraft.entity.attribute.EntityAttributes;
import net.minecraft.entity.mob.AbstractSkeletonEntity;
import net.minecraft.entity.mob.MobEntity;
import net.minecraft.entity.mob.ZombieEntity;
import net.minecraft.entity.projectile.ProjectileUtil;
import net.minecraft.item.ItemStack;
import net.minecraft.item.Items;
import net.minecraft.particle.ParticleTypes;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3d;
import org.jetbrains.annotations.Nullable;
import work.lclpnet.ap2.game.apocalypse_survival.goal.RoamGoal;
import work.lclpnet.ap2.game.apocalypse_survival.goal.UnstuckGoal;
import work.lclpnet.ap2.impl.util.WeightedList;
import work.lclpnet.ap2.impl.util.world.stage.Stage;
import work.lclpnet.ap2.mixin.MobEntityAccessor;
import work.lclpnet.kibu.scheduler.Ticks;

import java.util.Random;

public class MonsterSpawner {

    private static final int
            PARTICLE_TICKS = 12,
            MOB_MIN_TICKS = Ticks.seconds(1),
            MOB_MAX_TICKS = Ticks.seconds(3),
            MOB_LIMIT = 150;
    private final ServerWorld world;
    private final Stage stage;
    private final Random random;
    private final TargetManager targetManager;
    private final WeightedList<EntityType<? extends ZombieEntity>> zombieTypes;
    private final WeightedList<EntityType<? extends AbstractSkeletonEntity>> skeletonTypes;
    private final WeightedList<SpawnType> spawnTypes = new WeightedList<>();
    private int timeTicks = 0;
    private int nextParticle = 0;
    private int nextMob;
    private int mobCount = 0;

    public MonsterSpawner(ServerWorld world, Stage stage, Random random, TargetManager targetManager) {
        this.world = world;
        this.stage = stage;
        this.random = random;
        this.targetManager = targetManager;

        zombieTypes = new WeightedList<>();
        zombieTypes.add(EntityType.ZOMBIE, 0.8f);
        zombieTypes.add(EntityType.ZOMBIE_VILLAGER, 0.07f);
        zombieTypes.add(EntityType.HUSK, 0.05f);
        zombieTypes.add(EntityType.ZOMBIFIED_PIGLIN, 0.03f);
        zombieTypes.add(EntityType.DROWNED, 0.05f);

        skeletonTypes = new WeightedList<>();
        skeletonTypes.add(EntityType.SKELETON, 0.8f);
        skeletonTypes.add(EntityType.WITHER_SKELETON, 0.05f);
        skeletonTypes.add(EntityType.BOGGED, 0.05f);
        skeletonTypes.add(EntityType.STRAY, 0.08f);

        scheduleNextMob();
    }

    public void tick() {
        int time = timeTicks++;

        handleTimedEvents(time);

        if (nextParticle-- <= 0) {
            nextParticle = PARTICLE_TICKS;
            spawnParticle();
        }

        if (mobCount < MOB_LIMIT && nextMob-- <= 0) {
            scheduleNextMob();
            spawnMob();
        }
    }

    private void handleTimedEvents(int ticks) {
        switch (ticks) {
            case 0 -> spawnTypes.add(SpawnType.SKELETON, 0.6f);
            case 30 * 20 -> spawnTypes.add(SpawnType.SKELETON, 0.2f);
            default -> {}
        }
    }

    private void scheduleNextMob() {
        this.nextMob = MOB_MIN_TICKS + random.nextInt(MOB_MAX_TICKS - MOB_MIN_TICKS + 1);
    }

    private void spawnParticle() {
        BlockPos center = stage.getCenter();
        int offset = stage.getRadius() / 2;

        world.spawnParticles(ParticleTypes.REVERSE_PORTAL, center.getX() + 0.5, center.getY() + 0.5, center.getZ() + 0.5,
                30, offset, offset, offset, 0.15);
    }

    private void spawnMob() {
        SpawnType spawnType = spawnTypes.getRandomElement(random);

        switch (spawnType) {
            case ZOMBIE -> spawnZombie();
            case SKELETON -> spawnSkeleton();
            case null -> {}
        }
    }

    private void spawnZombie() {
        var zombie = createMob(zombieTypes);

        if (zombie == null) return;

        zombie.setCanBreakDoors(true);

        if (random.nextFloat() < 0.05f)  {
            zombie.setBaby(true);

            EntityAttributeInstance movementSpeed = zombie.getAttributeInstance(EntityAttributes.GENERIC_MOVEMENT_SPEED);

            if (movementSpeed != null) {
                movementSpeed.setBaseValue(0.18);
            }
        }

        // adjust goals
        var mobAccess = (MobEntityAccessor) zombie;

        GoalSelector goalSelector = mobAccess.getGoalSelector();

        GoalModifier.clear(goalSelector);
        GoalModifier.clear(mobAccess.getTargetSelector());

        goalSelector.add(1, new BreakDoorGoal(zombie, difficulty -> true));
        goalSelector.add(2, new ZombieAttackGoal(zombie, 1.5, false));
        goalSelector.add(7, new RoamGoal(zombie, targetManager, 1.25));
        goalSelector.add(8, new UnstuckGoal(zombie, random));

        // spawn mob
        spawnMobInWorld(zombie);
        targetManager.addZombie(zombie);
    }

    private void spawnSkeleton() {
        var skeleton = createMob(skeletonTypes);

        if (skeleton == null) return;

        // adjust goals
        var mobAccess = (MobEntityAccessor) skeleton;

        GoalSelector goalSelector = mobAccess.getGoalSelector();

        GoalModifier.clear(goalSelector);
        GoalModifier.clear(mobAccess.getTargetSelector());

        goalSelector.add(1, new BreakDoorGoal(skeleton, difficulty -> true));
        goalSelector.add(7, new RoamGoal(skeleton, targetManager, 1.25));
        goalSelector.add(8, new UnstuckGoal(skeleton, random));

        ItemStack stack = skeleton.getStackInHand(ProjectileUtil.getHandPossiblyHolding(skeleton, Items.BOW));

        if (stack.isOf(Items.BOW)) {
            goalSelector.add(4, new BowAttackGoal<>(skeleton, 1.2, 20, 15.0F));
        } else {
            goalSelector.add(4, new MeleeAttackGoal(skeleton, 1.5, false) {
                @Override
                public void start() {
                    super.start();
                    skeleton.setAttacking(true);
                }

                @Override
                public void stop() {
                    super.stop();
                    skeleton.setAttacking(false);
                }
            });
        }

        // spawn mob
        spawnMobInWorld(skeleton);
        targetManager.addSkeleton(skeleton);
    }

    @Nullable
    private <T extends MobEntity> T createMob(WeightedList<EntityType<? extends T>> types) {
        // select random zombie type
        var type = types.getRandomElement(random);

        if (type == null) return null;

        T mob = type.create(world, null, stage.getOrigin(), SpawnReason.COMMAND, false, false);

        if (mob == null) return null;

        configureMob(mob);

        return mob;
    }

    private void configureMob(MobEntity skeleton) {
        BlockPos pos = stage.getOrigin();

        skeleton.setPersistent();
        skeleton.setPosition(Vec3d.ofBottomCenter(pos));
        skeleton.setGlowing(true);  // debug

        // adjust follow range, so that the zombie will follow far players
        var followRange = skeleton.getAttributeInstance(EntityAttributes.GENERIC_FOLLOW_RANGE);

        if (followRange != null) {
            followRange.setBaseValue(100);
        }

        EntityNavigation navigation = skeleton.getNavigation();

        // adjust range multiplier to find longer paths using the A* Algorithm
        navigation.setRangeMultiplier(2.5f);

        if (navigation instanceof MobNavigation nav) {
            nav.setCanPathThroughDoors(true);
            nav.setCanWalkOverFences(true);
            nav.setCanEnterOpenDoors(true);
        }
    }

    private void spawnMobInWorld(MobEntity mob) {
        world.spawnEntity(mob);
        mobCount++;
    }

    private void debugPath(ZombieEntity zombie) {
        zombie.setOnGround(true);

        BlockPos target = new BlockPos(-50, 73, 10);

        EntityNavigation navigation = zombie.getNavigation();
        var path = navigation.findPathTo(target, 2, 150);

        new PathDebugging(world).displayPath(path, zombie);
    }

    private enum SpawnType {
        ZOMBIE,
        SKELETON,
    }
}

File: src/minigames/apocalypse_survival/src/main/kotlin/work/lclpnet/ap2/game/apocalypse_survival/util/MonsterSpawner.kt
package work.lclpnet.ap2.game.apocalypse_survival.util

import net.minecraft.core.BlockPos
import net.minecraft.core.particles.ParticleTypes
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.entity.EntitySpawnReason
import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.EntityTypes
import net.minecraft.world.entity.Mob
import net.minecraft.world.entity.ai.attributes.Attributes
import net.minecraft.world.entity.ai.goal.*
import net.minecraft.world.entity.ai.navigation.GroundPathNavigation
import net.minecraft.world.entity.monster.skeleton.AbstractSkeleton
import net.minecraft.world.entity.monster.zombie.Drowned
import net.minecraft.world.entity.monster.zombie.Zombie
import net.minecraft.world.entity.monster.zombie.ZombifiedPiglin
import net.minecraft.world.entity.projectile.ProjectileUtil
import net.minecraft.world.item.Items
import net.minecraft.world.phys.Vec3
import work.lclpnet.ap2.core.mixin.entity.MobAccessor
import work.lclpnet.ap2.ext.mc.isOf
import work.lclpnet.ap2.game.apocalypse_survival.goal.RoamGoal
import work.lclpnet.ap2.game.apocalypse_survival.goal.UnstuckGoal
import work.lclpnet.ap2.impl.util.GoalModifier
import work.lclpnet.ap2.impl.util.world.block_shape.BlockShape
import work.lclpnet.ap2.util.world.SizedSpaceFinder
import work.lclpnet.gaco.ds.WeightedList
import work.lclpnet.kibu.access.entity.EntityUtil
import java.util.*
import kotlin.math.pow

private const val PARTICLE_TICKS = 12
private const val MOB_LIMIT = 150

class MonsterSpawner(
    private val level: ServerLevel,
    private val stage: BlockShape,
    private val random: Random,
    private val targetManager: TargetManager,
    private val minSpawnTicks: Int,
    private val maxSpawnTicks: Int,
) {

    private val zombieTypes = WeightedList<EntityType<out Zombie>>().apply {
        addIfSupported(EntityTypes.ZOMBIE, 0.8f)
        addIfSupported(EntityTypes.ZOMBIE_VILLAGER, 0.07f)
        addIfSupported(EntityTypes.HUSK, 0.05f)
        addIfSupported(EntityTypes.ZOMBIFIED_PIGLIN, 0.03f)
        addIfSupported(EntityTypes.DROWNED, 0.05f)
    }
    private val skeletonTypes = WeightedList<EntityType<out AbstractSkeleton>>().apply {
        addIfSupported(EntityTypes.SKELETON, 0.8f)
        addIfSupported(EntityTypes.WITHER_SKELETON, 0.05f)
        addIfSupported(EntityTypes.BOGGED, 0.05f)
        addIfSupported(EntityTypes.STRAY, 0.08f)
    }
    private val spawnTypes = WeightedList<SpawnType>()
    private var timeTicks = 0
    private var nextParticle = 0
    private var nextMob = scheduleNextMob()
    private var mobCount = 0

    fun tick() {
        val time = timeTicks++

        handleTimedEvents(time)

        if (nextParticle-- <= 0) {
            nextParticle = PARTICLE_TICKS
            spawnParticle()
        }

        if (mobCount < MOB_LIMIT && nextMob-- <= 0) {
            nextMob = scheduleNextMob()
            spawnMob()
        }
    }

    private fun handleTimedEvents(ticks: Int) {
        when (ticks) {
            0 -> addSpawnType(SpawnType.ZOMBIE, 0.6f)
            70 * 20 -> addSpawnType(SpawnType.VINDICATOR, 0.06f)
            130 * 20 -> addSpawnType(SpawnType.SKELETON, 0.165f)
            150 * 20 -> addSpawnType(SpawnType.GHAST, 0.05f)
            175 * 20 -> addSpawnType(SpawnType.EVOKER, 0.06f)
            200 * 20 -> addSpawnType(SpawnType.PHANTOM, 0.04f)
        }
    }

    private fun addSpawnType(type: SpawnType, weight: Float) {
        val supported = when (type) {
            SpawnType.ZOMBIE -> zombieTypes.isNotEmpty()
            SpawnType.SKELETON -> skeletonTypes.isNotEmpty()
            SpawnType.PHANTOM -> supports(EntityTypes.PHANTOM)
            SpawnType.GHAST -> supports(EntityTypes.GHAST)
            SpawnType.VINDICATOR -> supports(EntityTypes.VINDICATOR)
            SpawnType.EVOKER -> supports(EntityTypes.EVOKER)
        }

        if (supported) {
            spawnTypes.add(type, weight)
        }
    }

    private fun scheduleNextMob() =
        minSpawnTicks + random.nextInt(maxSpawnTicks - minSpawnTicks + 1)

    private fun spawnParticle() {
        if (stage !is BlockShape.WithRadius) return

        val center: BlockPos = stage.center()
        val offset = stage.radius() / 2

        level.sendParticles(
            ParticleTypes.REVERSE_PORTAL,
            center.x + 0.5, center.y + 0.5, center.z + 0.5,
            30, offset.toDouble(), offset.toDouble(), offset.toDouble(), 0.15
        )
    }

    private fun spawnMob() {
        when (spawnTypes.getRandomElement(random)) {
            SpawnType.ZOMBIE -> spawnZombie()
            SpawnType.SKELETON -> spawnSkeleton()
            SpawnType.PHANTOM -> spawnPhantom()
            SpawnType.GHAST -> spawnGhast()
            SpawnType.VINDICATOR -> spawnVindicator()
            SpawnType.EVOKER -> spawnEvoker()
            null -> {}
        }
    }

    private fun spawnZombie() {
        val zombie = createMob(zombieTypes) ?: return

        zombie.setCanBreakDoors(true)

        var baseSpeed = zombie.getAttributeBaseValue(Attributes.MOVEMENT_SPEED)

        if (zombie.isBaby) {
            baseSpeed *= 0.75
        } else if (random.nextFloat() < 0.05f) {
            val scale = random.nextDouble(0.75, 1.8)

            if (supports(zombie.type, scale)) {
                EntityUtil.setAttribute(zombie, Attributes.SCALE, scale)
                baseSpeed *= (1.0 / scale.pow(1.15)).coerceIn(0.75, 1.12)
            }
        }

        if (zombie is Drowned) {
            baseSpeed *= 0.87
        } else if (zombie is ZombifiedPiglin) {
            baseSpeed *= 0.9
        }

        EntityUtil.setAttribute(zombie, Attributes.MOVEMENT_SPEED, baseSpeed)

        val mobAccess = zombie as MobAccessor
        val goalSelector: GoalSelector = mobAccess.goalSelector

        GoalModifier.clear(goalSelector)
        GoalModifier.clear(mobAccess.targetSelector)

        goalSelector.addGoal(1, BreakDoorGoal(zombie) { true })
        goalSelector.addGoal(2, ZombieAttackGoal(zombie, 1.4, false))
        goalSelector.addGoal(7, RoamGoal(zombie, targetManager, 1.25))
        goalSelector.addGoal(8, UnstuckGoal(zombie, random))

        if (zombie is Drowned) {
            goalSelector.addGoal(2, Drowned.DrownedTridentAttackGoal(zombie, 1.0, 60, 10.0f))
        }

        spawnMobInWorld(zombie)
    }

    private fun spawnSkeleton() {
        val skeleton = createMob(skeletonTypes) ?: return

        var baseSpeed = skeleton.getAttributeBaseValue(Attributes.MOVEMENT_SPEED)

        if (random.nextFloat() < 0.075f) {
            val scale = random.nextDouble(0.75, 2.5)

            if (supports(skeleton.type, scale)) {
                EntityUtil.setAttribute(skeleton, Attributes.SCALE, scale)
                baseSpeed *= (1.0 / scale.pow(1.15)).coerceIn(0.6, 1.1)
            }
        }

        EntityUtil.setAttribute(skeleton, Attributes.MOVEMENT_SPEED, baseSpeed)

        val mobAccess = skeleton as MobAccessor
        val goalSelector: GoalSelector = mobAccess.goalSelector

        GoalModifier.clear(goalSelector)
        GoalModifier.clear(mobAccess.targetSelector)

        goalSelector.addGoal(1, BreakDoorGoal(skeleton) { true })
        goalSelector.addGoal(7, RoamGoal(skeleton, targetManager, 1.25))
        goalSelector.addGoal(8, UnstuckGoal(skeleton, random))

        val stack = skeleton.getItemInHand(ProjectileUtil.getWeaponHoldingHand(skeleton, Items.BOW))

        if (stack.isOf(Items.BOW)) {
            goalSelector.addGoal(4, RangedBowAttackGoal(skeleton, 1.1, 20, 15.0f))
        } else {
            goalSelector.addGoal(4, object : MeleeAttackGoal(skeleton, 1.4, false) {
                override fun start() {
                    super.start()
                    skeleton.isAggressive = true
                }

                override fun stop() {
                    super.stop()
                    skeleton.isAggressive = false
                }
            })
        }

        spawnMobInWorld(skeleton)
    }

    private fun spawnPhantom() {
        val phantom = createMob(EntityTypes.PHANTOM) ?: return

        phantom.phantomSize = 0

        if (random.nextFloat() < 0.25f) {
            val scale = random.nextDouble(0.2, 5.0)

            if (supports(phantom.type, scale)) {
                EntityUtil.setAttribute(phantom, Attributes.SCALE, scale)
            }
        }

        val mobAccess = phantom as MobAccessor
        GoalModifier.clear(mobAccess.targetSelector)

        spawnMobInWorld(phantom)
    }

    private fun spawnGhast() {
        val ghast = createMob(EntityTypes.GHAST) ?: return

        val scale = random.nextDouble(0.2, 1.0)

        if (supports(ghast.type, scale)) {
            EntityUtil.setAttribute(ghast, Attributes.SCALE, scale)
        }

        spawnMobInWorld(ghast)
    }

    private fun spawnVindicator() {
        val vindicator = createMob(EntityTypes.VINDICATOR) ?: return

        var baseSpeed = vindicator.getAttributeBaseValue(Attributes.MOVEMENT_SPEED)

        if (random.nextFloat() < 0.05f) {
            val scale = random.nextDouble(0.75, 1.3)

            if (supports(vindicator.type, scale)) {
                EntityUtil.setAttribute(vindicator, Attributes.SCALE, scale)
                baseSpeed *= (1.0 / scale.pow(1.15)).coerceIn(0.75, 1.12)
            }
        }

        EntityUtil.setAttribute(vindicator, Attributes.MOVEMENT_SPEED, baseSpeed)

        spawnMobInWorld(vindicator)
    }

    private fun spawnEvoker() {
        val evoker = createMob(EntityTypes.EVOKER) ?: return
        spawnMobInWorld(evoker)
    }

    private fun <T : Mob> createMob(types: WeightedList<EntityType<out T>>): T? {
        val type = types.getRandomElement(random) ?: return null
        return createMob(type)
    }

    private fun <T : Mob> createMob(type: EntityType<out T>): T? {
        if (!supports(type)) return null

        val mob = type.create(
            level,
            null,
            stage.origin(),
            EntitySpawnReason.COMMAND,
            false,
            false
        ) ?: return null

        configureMob(mob)

        return mob
    }

    fun supports(type: EntityType<*>, scale: Double = 1.0): Boolean {
        val dimensions = type.dimensions

        val width = dimensions.width().toDouble() * scale
        val height = dimensions.height().toDouble() * scale

        val pos = stage.origin()

        return SizedSpaceFinder.hasSpaceAt(
            level,
            pos.x + 0.5,
            pos.y.toDouble(),
            pos.z + 0.5,
            width,
            height,
            width
        )
    }

    private fun configureMob(mob: Mob) {
        val pos: BlockPos = stage.origin()

        mob.setPersistenceRequired()
        mob.setPos(Vec3.atBottomCenterOf(pos))

        mob.getAttribute(Attributes.FOLLOW_RANGE)?.baseValue = 100.0

        val navigation = mob.navigation
        navigation.setMaxVisitedNodesMultiplier(2.5f)

        if (navigation is GroundPathNavigation) {
            navigation.setCanOpenDoors(true)
            navigation.setCanWalkOverFences(true)
        }
    }

    private fun spawnMobInWorld(mob: Mob) {
        level.addFreshEntity(mob)
        mobCount++
    }

    private enum class SpawnType {
        ZOMBIE, SKELETON, PHANTOM, GHAST, VINDICATOR, EVOKER
    }

    private fun <T : EntityType<*>> WeightedList<T>.addIfSupported(type: T, weight: Float) {
        if (supports(type)) {
            add(type, weight)
        }
    }
}
